File: sdk-go/client_test.go
package ztxbas

import (
	"context"
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"net/http/httptest"
	"strconv"
	"testing"
	"time"
)

// verifySig re-runs the server's HMAC check inside the test server so we
// can be certain the SDK is producing headers the real server would
// accept. Any deviation shows up here.
func verifySig(t *testing.T, r *http.Request, secret string) {
	t.Helper()
	body, _ := io.ReadAll(r.Body)
	r.Body = io.NopCloser(bytesReader(body))

	ts := r.Header.Get(hdrTimestamp)
	nonce := r.Header.Get(hdrNonce)
	sig := r.Header.Get(hdrSignature)
	appID := r.Header.Get(hdrApplicationID)
	if appID == "" || ts == "" || nonce == "" || sig == "" {
		t.Fatalf("missing HMAC headers: app=%q ts=%q nonce=%q sig=%q", appID, ts, nonce, sig)
	}
	if len(sig) != 64 {
		t.Fatalf("signature length %d, want 64", len(sig))
	}
	if _, err := strconv.ParseInt(ts, 10, 64); err != nil {
		t.Fatalf("timestamp not integer: %v", err)
	}
	canonical := fmt.Sprintf("%s|%s|%s|%s|%s", r.Method, r.URL.Path, ts, nonce, body)
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write([]byte(canonical))
	want := hex.EncodeToString(mac.Sum(nil))
	if want != sig {
		t.Fatalf("signature mismatch\n canonical: %s\n want: %s\n got:  %s", canonical, want, sig)
	}
}

// bytesReader avoids pulling bytes into imports; tests are cleaner with
// this helper than an extra import block per file.
func bytesReader(b []byte) io.Reader {
	return &readerFromBytes{b: b}
}

type readerFromBytes struct {
	b []byte
	i int
}

func (r *readerFromBytes) Read(p []byte) (int, error) {
	if r.i >= len(r.b) {
		return 0, io.EOF
	}
	n := copy(p, r.b[r.i:])
	r.i += n
	return n, nil
}

func newTestClient(t *testing.T, handler http.Handler) (*Client, *httptest.Server) {
	t.Helper()
	ts := httptest.NewServer(handler)
	t.Cleanup(ts.Close)
	c, err := New(ts.URL, "app_test", "secret_test")
	if err != nil {
		t.Fatalf("New: %v", err)
	}
	return c, ts
}

func TestClient_RegisterUser(t *testing.T) {
	c, _ := newTestClient(t, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v1/users" || r.Method != "POST" {
			t.Fatalf("unexpected %s %s", r.Method, r.URL.Path)
		}
		verifySig(t, r, "secret_test")
		if r.Header.Get("Content-Type") != "application/json" {
			t.Errorf("missing Content-Type")
		}
		w.WriteHeader(http.StatusCreated)
		_, _ = w.Write([]byte(`{"id":"u_1","email":"alice@example.com","enrolled":false}`))
	}))

	u, err := c.RegisterUser(context.Background(), RegisterUserRequest{Email: "alice@example.com"})
	if err != nil {
		t.Fatalf("RegisterUser: %v", err)
	}
	if u.ID != "u_1" || u.Email != "alice@example.com" || u.Enrolled {
		t.Errorf("unexpected user %+v", u)
	}
}

func TestClient_ListUsers(t *testing.T) {
	c, _ := newTestClient(t, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v1/users" || r.Method != "GET" {
			t.Fatalf("unexpected %s %s", r.Method, r.URL.Path)
		}
		verifySig(t, r, "secret_test")
		_, _ = w.Write([]byte(`{"users":[{"id":"u_1","email":"a@x","enrolled":true}]}`))
	}))
	list, err := c.ListUsers(context.Background())
	if err != nil {
		t.Fatalf("ListUsers: %v", err)
	}
	if len(list) != 1 || list[0].Email != "a@x" {
		t.Errorf("unexpected list: %+v", list)
	}
}

func TestClient_DeregisterUser_NoContent(t *testing.T) {
	c, _ := newTestClient(t, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		verifySig(t, r, "secret_test")
		w.WriteHeader(http.StatusNoContent)
	}))
	if err := c.DeregisterUser(context.Background(), "alice@example.com"); err != nil {
		t.Errorf("DeregisterUser: %v", err)
	}
}

func TestClient_RegisterOrigin(t *testing.T) {
	c, _ := newTestClient(t, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		verifySig(t, r, "secret_test")
		var body map[string]any
		_ = json.NewDecoder(r.Body).Decode(&body)
		if body["origin"] != "https://app.example.com" {
			t.Errorf("body: %v", body)
		}
		w.WriteHeader(http.StatusCreated)
		_, _ = w.Write([]byte(`{"id":"o_1","origin":"https://app.example.com","origin_hash":"abc","display_name":"App"}`))
	}))
	o, err := c.RegisterOrigin(context.Background(), RegisterOriginRequest{
		Origin: "https://app.example.com", DisplayName: "App",
	})
	if err != nil {
		t.Fatalf("RegisterOrigin: %v", err)
	}
	if o.ID != "o_1" {
		t.Errorf("id: %+v", o)
	}
}

func TestClient_DeleteOrigin_PathEscape(t *testing.T) {
	var seenPath string
	c, _ := newTestClient(t, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		verifySig(t, r, "secret_test")
		seenPath = r.URL.Path
		w.WriteHeader(http.StatusNoContent)
	}))
	if err := c.DeleteOrigin(context.Background(), "o_1"); err != nil {
		t.Fatalf("DeleteOrigin: %v", err)
	}
	if seenPath != "/v1/origins/o_1" {
		t.Errorf("path: %q", seenPath)
	}
}

func TestClient_CreateChallenge(t *testing.T) {
	c, _ := newTestClient(t, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		verifySig(t, r, "secret_test")
		w.WriteHeader(http.StatusCreated)
		_, _ = w.Write([]byte(`{"challenge_id":"c_1","expires_in":300,"origin_display":"App","origin_url":"https://app.example.com"}`))
	}))
	ch, err := c.CreateChallenge(context.Background(), CreateChallengeRequest{
		UserEmail: "a@x", Origin: "https://app.example.com",
	})
	if err != nil {
		t.Fatalf("CreateChallenge: %v", err)
	}
	if ch.ChallengeID != "c_1" || ch.ExpiresIn != 300 {
		t.Errorf("resp: %+v", ch)
	}
}

func TestClient_ErrorMapping(t *testing.T) {
	tests := []struct {
		status  int
		code    string
		wantErr error
	}{
		{http.StatusUnauthorized, "INVALID_SIGNATURE", ErrUnauthorized},
		{http.StatusNotFound, "USER_NOT_FOUND", ErrNotFound},
		{http.StatusConflict, "DUPLICATE_EMAIL", ErrConflict},
		{http.StatusForbidden, "UNREGISTERED_ORIGIN", ErrUnregisteredOrigin},
	}
	for _, tc := range tests {
		t.Run(tc.code, func(t *testing.T) {
			c, _ := newTestClient(t, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
				w.WriteHeader(tc.status)
				fmt.Fprintf(w, `{"error":%q,"message":"boom"}`, tc.code)
			}))
			_, err := c.ListUsers(context.Background())
			if !errors.Is(err, tc.wantErr) {
				t.Errorf("errors.Is(_, %v) = false; got %v", tc.wantErr, err)
			}
			var apiErr *APIError
			if !errors.As(err, &apiErr) {
				t.Errorf("errors.As APIError = false; got %v", err)
			} else if apiErr.Code != tc.code {
				t.Errorf("code %q, want %q", apiErr.Code, tc.code)
			}
		})
	}
}

func TestClient_TimestampWithinSkew(t *testing.T) {
	c, _ := newTestClient(t, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		ts, _ := strconv.ParseInt(r.Header.Get(hdrTimestamp), 10, 64)
		if diff := time.Since(time.Unix(ts, 0)); diff < 0 || diff > 5*time.Second {
			t.Errorf("timestamp drift %v", diff)
		}
		w.WriteHeader(http.StatusNoContent)
	}))
	_ = c.DeregisterUser(context.Background(), "a@x")
}

func TestNew_Validation(t *testing.T) {
	if _, err := New("", "a", "s"); err == nil {
		t.Error("empty baseURL should error")
	}
	if _, err := New("https://x", "", "s"); err == nil {
		t.Error("empty appID should error")
	}
	if _, err := New("https://x", "a", ""); err == nil {
		t.Error("empty secret should error")
	}
	if _, err := New("not-a-url", "a", "s"); err == nil {
		t.Error("bare host without scheme should error")
	}
}

File: sdk-node/test/client.test.ts
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { createServer, Server, IncomingMessage, ServerResponse } from 'node:http';
import { createHmac } from 'node:crypto';
import {
  Client,
  ConflictError,
  NotFoundError,
  UnauthorizedError,
  UnregisteredOriginError,
  ApiError,
} from '../src/index.js';

/**
 * Spin up a real HTTP server on an ephemeral port, hand its base URL to a
 * Client, and return both. The handler receives the request plus its
 * fully-read body; test bodies can inspect and reply however they like.
 */
async function withServer(
  handler: (req: IncomingMessage, res: ServerResponse, body: string) => void,
): Promise<{ client: Client; server: Server; url: string }> {
  const server = createServer((req, res) => {
    const chunks: Buffer[] = [];
    req.on('data', (c) => chunks.push(c));
    req.on('end', () => handler(req, res, Buffer.concat(chunks).toString('utf8')));
  });
  await new Promise<void>((resolve) => server.listen(0, '127.0.0.1', resolve));
  const addr = server.address();
  if (typeof addr === 'string' || !addr) throw new Error('unexpected address');
  const url = `http://127.0.0.1:${addr.port}`;
  const client = new Client(url, 'app_test', 'secret_test');
  return { client, server, url };
}

/**
 * Re-run the server's HMAC check on an inbound request. If this fails,
 * the SDK produced headers the real server would reject — the whole
 * point of pinning tests here rather than only in a unit test.
 */
function assertSigValid(req: IncomingMessage, body: string, secret: string) {
  const appId = req.headers['x-application-id'] as string;
  const ts = req.headers['x-timestamp'] as string;
  const nonce = req.headers['x-nonce'] as string;
  const sig = req.headers['x-signature'] as string;
  assert.ok(appId && ts && nonce && sig, 'missing HMAC headers');
  assert.equal(sig.length, 64);
  const canonical = `${req.method}|${req.url}|${ts}|${nonce}|${body}`;
  const want = createHmac('sha256', secret).update(canonical).digest('hex');
  assert.equal(sig, want, 'signature mismatch');
}

test('registerUser sends signed POST and parses response', async () => {
  const { client, server } = await withServer((req, res, body) => {
    assert.equal(req.method, 'POST');
    assert.equal(req.url, '/v1/users');
    assertSigValid(req, body, 'secret_test');
    res.writeHead(201, { 'Content-Type': 'application/json' });
    res.end('{"id":"u_1","email":"alice@example.com","enrolled":false}');
  });
  try {
    const u = await client.registerUser({ email: 'alice@example.com' });
    assert.equal(u.id, 'u_1');
    assert.equal(u.email, 'alice@example.com');
    assert.equal(u.enrolled, false);
  } finally {
    server.close();
  }
});

test('listUsers unwraps the {users:[…]} envelope', async () => {
  const { client, server } = await withServer((req, res) => {
    res.writeHead(200, { 'Content-Type': 'application/json' });
    res.end('{"users":[{"id":"u_1","email":"a@x","enrolled":true}]}');
  });
  try {
    const list = await client.listUsers();
    assert.equal(list.length, 1);
    assert.equal(list[0].email, 'a@x');
  } finally {
    server.close();
  }
});

test('deregisterUser handles 204', async () => {
  const { client, server } = await withServer((req, res, body) => {
    assertSigValid(req, body, 'secret_test');
    assert.equal(req.method, 'DELETE');
    res.writeHead(204);
    res.end();
  });
  try {
    await client.deregisterUser('alice@example.com');
  } finally {
    server.close();
  }
});

test('registerOrigin round trips', async () => {
  const { client, server } = await withServer((req, res, body) => {
    assertSigValid(req, body, 'secret_test');
    const parsed = JSON.parse(body);
    assert.equal(parsed.origin, 'https://app.example.com');
    assert.equal(parsed.display_name, 'App');
    res.writeHead(201, { 'Content-Type': 'application/json' });
    res.end('{"id":"o_1","origin":"https://app.example.com","origin_hash":"abc","display_name":"App"}');
  });
  try {
    const o = await client.registerOrigin({
      origin: 'https://app.example.com',
      display_name: 'App',
    });
    assert.equal(o.id, 'o_1');
  } finally {
    server.close();
  }
});

test('deleteOrigin uses percent-encoded path segment', async () => {
  let seenPath = '';
  const { client, server } = await withServer((req, res) => {
    seenPath = req.url ?? '';
    res.writeHead(204);
    res.end();
  });
  try {
    await client.deleteOrigin('o_1');
    assert.equal(seenPath, '/v1/origins/o_1');
  } finally {
    server.close();
  }
});

test('createChallenge parses full response', async () => {
  const { client, server } = await withServer((req, res, body) => {
    assertSigValid(req, body, 'secret_test');
    res.writeHead(201, { 'Content-Type': 'application/json' });
    res.end(
      '{"challenge_id":"c_1","expires_in":300,"origin_display":"App","origin_url":"https://app.example.com"}',
    );
  });
  try {
    const ch = await client.createChallenge({
      user_email: 'a@x',
      origin: 'https://app.example.com',
    });
    assert.equal(ch.challenge_id, 'c_1');
    assert.equal(ch.expires_in, 300);
  } finally {
    server.close();
  }
});

test('error responses map to typed subclasses', async () => {
  const cases: Array<{ status: number; code: string; ctor: unknown }> = [
    { status: 401, code: 'INVALID_SIGNATURE', ctor: UnauthorizedError },
    { status: 404, code: 'USER_NOT_FOUND', ctor: NotFoundError },
    { status: 409, code: 'DUPLICATE_EMAIL', ctor: ConflictError },
    { status: 403, code: 'UNREGISTERED_ORIGIN', ctor: UnregisteredOriginError },
    { status: 500, code: 'INTERNAL', ctor: ApiError },
  ];
  for (const c of cases) {
    const { client, server } = await withServer((req, res) => {
      res.writeHead(c.status, { 'Content-Type': 'application/json' });
      res.end(JSON.stringify({ error: c.code, message: 'boom' }));
    });
    try {
      let caught: unknown;
      try {
        await client.listUsers();
      } catch (e) {
        caught = e;
      }
      assert.ok(caught instanceof (c.ctor as { new (...args: never[]): Error }), `expected ${(c.ctor as { name: string }).name} for ${c.code}, got ${caught}`);
      const apiErr = caught as ApiError;
      assert.equal(apiErr.code, c.code);
      assert.equal(apiErr.statusCode, c.status);
    } finally {
      server.close();
    }
  }
});

test('constructor rejects missing arguments', () => {
  assert.throws(() => new Client('', 'a', 's'));
  assert.throws(() => new Client('https://x', '', 's'));
  assert.throws(() => new Client('https://x', 'a', ''));
  assert.throws(() => new Client('not-a-url', 'a', 's'));
});

test('constructor strips trailing slash from baseUrl', async () => {
  let seenPath = '';
  const server = createServer((req, res) => {
    seenPath = req.url ?? '';
    res.writeHead(204);
    res.end();
  });
  await new Promise<void>((resolve) => server.listen(0, '127.0.0.1', resolve));
  const addr = server.address();
  const port = typeof addr === 'object' && addr ? addr.port : 0;
  try {
    const c = new Client(`http://127.0.0.1:${port}/`, 'a', 's');
    await c.deregisterUser('x@y');
    // Should be exactly one leading slash, not two.
    assert.equal(seenPath, '/v1/users');
  } finally {
    server.close();
  }
});
